tests/testthat/helper-shinytest2.R

```diff
@@ -0,0 +1,254 @@
+
+# get_module_input(app, "dataname")
+get_module_input <- function(app, input_id) {
+  active_ns <- get_active_ns(app, "module")
+  app$get_value(input = sprintf("%s-%s", active_ns, input_id))
+}
+
+# get_module_output(app, "text")
+get_module_output <- function(app, output_id) {
+  active_ns <- get_active_ns(app, "module")
+  app$get_value(output = sprintf("%s-%s", active_ns, output_id))
+}
+
+
+# get_app_modules(app)
+get_app_module_tabs <- function(app) {
+  lapply(
+    app$get_html(selector = "ul.shiny-bound-input"),
+    function(x) {
+      el <- rvest::read_html(x)
+      root_id <- el %>%
+        html_node("ul") %>%
+        html_attr("id") %>%
+        gsub(pattern = "(^teal-main_ui-)|(-active_tab$)", replacement = "")
+      tab_id <- el %>%
+        html_nodes("li a") %>%
+        html_attr("data-value")
+      tab_name <- el %>%
+        html_nodes("li a") %>%
+        html_text()
+      data.frame(
+        root_id = root_id,
+        tab_id = tab_id,
+        tab_name = tab_name
+      )
+    }
+  ) %>%
+    do.call(what = rbind)
+}
+
+# get_active_filter_vars(app)
+get_active_filter_vars <- function(app) {
+  displayed_data_index <- sapply(
+    app$get_html(
+      sprintf(
+        "#%s-active-filter_active_vars_contents > span",
+        get_active_ns(app, "filter")
+      )
+    ),
+    function(x) {
+      style <- x %>%
+        rvest::read_html() %>%
+        rvest::html_node("span") %>%
+        rvest::html_attr("style")
+      style <- ifelse(is.na(style), "", style)
+      style != "display: none;"
+    }
+  ) %>%
+    unname()
+  available_data <- app$get_html(
+    sprintf(
+      "#%s-active-filter_active_vars_contents",
+      get_active_ns(app, "filter")
+    )
+  ) %>%
+    read_html() %>%
+    html_nodes(".filter_panel_dataname") %>%
+    html_text()
+  available_data[displayed_data_index]
+}
+
+# get_active_data_filters(app, "mtcars")
+get_active_data_filters <- function(app, data_name) {
+  sapply(
+    app$get_html(
+      sprintf(
+        "#%s-active-%s-filter-cards .filter-card-varname",
+        get_active_ns(app, "filter"),
+        data_name
+      )
+    ),
+    function(x) {
+      x %>%
+        rvest::read_html() %>%
+        rvest::html_text() %>%
+        gsub(pattern = "\\s", replacement = "")
+    }
+  ) %>%
+    unname()
+}
+
+# get_active_selection_value(app, "iris", "Species")
+get_active_selection_value <- function(app, data_name, filter_name, is_numeric = FALSE) {
+  selection_suffix <- ifelse(is_numeric, "selection_manual", "selection")
+  app$get_value(
+    input = sprintf(
+      "%s-active-%s-filter-%s_%s-inputs-%s",
+      get_active_ns(app, "filter"),
+      data_name,
+      data_name,
+      filter_name,
+      selection_suffix
+    )
+  )
+}
+
+# add_filter_var(app, "mtcars", "wt")
+add_filter_var <- function(app, data_name, var_name) {
+  app$set_inputs(
+    !!sprintf(
+      "%s-add-%s-filter-var_to_add",
+      get_active_ns(app, "filter"),
+      data_name
+    ) := var_name
```

Comment:
If there is any way to omit `!!` and `:=` then I would appreciate it : D

Reply:
I guess it's unavoidable

Reply:
I tried to eliminate it for a while and was unsuccessful.

Reply:
Well now I know :)